tools/c7n_gcp/c7n_gcp/query.py

```diff
@@ -145,7 +145,45 @@ def get_resource_query(self):
             return {'filter': self.data.get('query')}
 
     def resources(self, query=None):
-        q = query or self.get_resource_query()
+        if hasattr(self.resource_type, 'parent'):
+            return self._fetch_resources_relationally(query)
+
+        return self._fetch_resources_directly(query)
+
+    def _fetch_resources_relationally(self, query):
+        """ Fetches resources by first fetching their parents and extracting
+        required arguments from the latter ones. Injects each corresponding
+        parent into a child resource afterwards.
+        """
+
+        resources = []
+
+        parent_description = self.resource_type.parent
+        resource_type, args_extractor = parent_description[:-1]
+
+        injection_name = 'c7n:{}'.format(
```

Comment:
we use annotation key, and in this case it should probably just be static like c7n:parent... the variable use on the conditional set looks pretty odd, what's the intent there?

Reply:
I didn't use static annotation on purpose. When your level of nesting is more than one (e.g. in KMS case: locations -> KeyRings -> CryptoKeys -> etc), you can clearly see the whole chain.
E.g.  `c7n:cryptokey.c7n:keyring.c7n:location` vs `c7n:parent.c7n:parent.c7n:parent`

Reply:
ic. that seems reasonable, although having commonality around parent also seems nice. the logic for the set here could be simplified by just using `parent_mgr.type` instead of magic offsets

Reply:
> ic. that seems reasonable, although having commonality around parent also seems nice. the logic for the set here could be simplified by just using `parent_mgr.type` instead of magic offsets

Not sure I got you correctly. Do you mean using parent's `resource_type` without providing possibility to use custom annotation? This was introduced for those resources prefixed with a service name. E.g. for Cloud SQL we have `sql-instance` resource and if we work with SQL database resource (see example in the descr), we already know that we are in the context of SQL, so there is no need of extra prefix `sql-`. Thus we can change annotation (injection name) to be `instance` instead (`c7n:instance` vs `c7n:sql-instance`)

Reply:
resource_manager.type is a string for the policy resource, which is what was trying to be extracted out out of the type info's parent spec. rather than doing len() == 3 etc.. its much cleaner to just resource_manager.type in the format string. that will still have the parent's type in the annotation key.

Reply:
The reason of doing `len() == 3` is optional parameter which, if presented, is meant to not be equal to the resource type (my previous msg).
If we decided to not provide the possibility, then I would write
```python
'c7n:{}'.format(resource_type)
```
and there would be no need for doing `resource_manager.type` because I already have that information extracted in the following line (know you noticed bug there):
```python
resource_type, args_extractor = parent_description[:-1]
```

Reply:
So considering the example given above (`c7n:instance` vs `c7n:sql-instance`), would you like me to get rid off that possibility for giving a custom injection name (always default to `resource_type`)?

Reply:
Deleted the possibility but will remain the conversation opened in case you have some additional notes.

Reply:
yeah, standardizing on parent type is good, no need for customization here that i can see. thanks.